doc/manual/metaprogramming.rst

```diff
@@ -872,3 +872,102 @@ entirely in Julia. You can read their source and see precisely what they
 do — and all they do is construct expression objects to be inserted into
 your program's syntax tree.
 
+Staged functions
+----------------
+
+Sometimes, generating or manipulating expressions through macros is a little
+overkill for the problem you're trying to solve, but you still need some
+more advanced magic than just regular functions; enter *staged functions*.
+Staged functions have the capability to generate specialized code depending
+on the *types* of the arguments you give them, so that you can optimize or
+generalize your code in ways that aren't possible with ordinary functions.
```

Comment:
A couple of points:
- This description makes it sound as if stagedfunctions are a little less powerful than macros, but mostly they are just different (and there are a good number of things you can do with stagedfunctions that you can't do with macros). The key distinctions between macros and stagedfunctions are:

  + macros only work with expressions, and so don't know the types of the inputs
  + macros work at parsing-time, whereas stagedfunctions get expanded on-demand, possibly differently for each set of types

- One way I like to conceptualize stagedfunctions is that they provide a flexible framework to move work from runtime to compile-time. I'll see if I can come up with a short example below.